llvm/lib/Target/AIE/ReservedRegsLICM.cpp

```diff
@@ -0,0 +1,330 @@
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+private:
+  /// Keep track of information about hoisting candidates.
+  struct CandidateInfo {
+    CandidateInfo(MCPhysReg DefinedReg) : DefinedReg(DefinedReg) {}
+    MCPhysReg DefinedReg = MCRegister::NoRegister;
+    MachineInstr *HoistCandidate = nullptr;
+  };
+  DenseMap<MCPhysReg, CandidateInfo> Candidates;
+  CandidateInfo *getCandidateInfo(const MachineInstr &MI);
+
+  /// Collect all simplifiable reserved registers that all livein for \p L.
+  BitVector collectLoopReservedLiveins(const MachineLoop &L);
+
+  /// Go through \p L to look for instructions to hoist in the preheader or
+  /// exit block.
+  void runOnLoop(MachineLoop &L);
+
+  /// Find instructions to hoist to the exit block
+  void processForExitHoist(MachineLoop &L, const BitVector &ReservedLiveins);
+
+  /// Hoist \p Cand to \p L's exit block it it is safe to do so.
+  bool tryHoistToExitBlock(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Hoist \p Cand to \p L's preheader it it is safe to do so.
+  bool tryHoistToPreHeader(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Hoist \p Cand out of the loop to \p InsertBefore.
+  void hoist(const CandidateInfo &Cand,
+             MachineBasicBlock::iterator InsertBefore,
+             MachineBasicBlock &InsertMBB);
+
+  /// Verify if \p Cand is loop invariant and can be safely hoisted.
+  /// \pre Cand->DefinedReg has a unique live value within the loop. This is
+  /// verified by processForExitHoist().
+  bool isLoopInvariantInst(const CandidateInfo &Cand, const MachineLoop &L);
+};
+
+} // end anonymous namespace
+
+char ReservedRegsLICM::ID;
+
+char &llvm::ReservedRegsLICMID = ReservedRegsLICM::ID;
+
+INITIALIZE_PASS_BEGIN(ReservedRegsLICM, DEBUG_TYPE,
+                      "Machine LICM for reserved regs", false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineLoopInfo)
+INITIALIZE_PASS_END(ReservedRegsLICM, DEBUG_TYPE,
+                    "Machine LICM for reserved regs", false, false)
+
+llvm::FunctionPass *llvm::createReservedRegsLICMPass() {
+  return new ReservedRegsLICM();
+}
+
+bool ReservedRegsLICM::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction()))
+    return false;
+
+  const TargetSubtargetInfo &ST = MF.getSubtarget();
+  TRI = ST.getRegisterInfo();
+  MRI = &MF.getRegInfo();
+  assert(MRI->isSSA() && "PhysRegCopyTracker can only track SSA copies");
+
+  LLVM_DEBUG(dbgs() << "******** Reserved register LICM: " << MF.getName()
+                    << " ********\n");
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  for (MachineLoop *L : MLI.getBase()) {
+    runOnLoop(*L);
+  }
+
+  return Changed;
+}
+
+static MCRegister getSinglePhysRegDef(const MachineInstr &MI) {
+  if (MI.getNumDefs() == 1 && MI.getNumExplicitDefs() == 1) {
+    const MachineOperand &MO = MI.getOperand(0);
+    return MO.isReg() && MO.getReg().isPhysical() ? MO.getReg().asMCReg()
```

Comment:
Is there any possibility that MO is not a register? I think your first If guarantees this condition.

I think it that it can be simpler:
```
if (MO.getReg().isPhysical()) 
    return MO.getReg().asMCReg();
```